File: apps/android/app/src/test/java/dev/tohure/android_rust_test/adapter/FakeCoreFinanciero.kt
package dev.tohure.android_rust_test.adapter

import uniffi.core_financiero.Account
import uniffi.core_financiero.TransferRequest
import uniffi.core_financiero.TransferResult
import uniffi.core_financiero.ValidCard
import uniffi.core_financiero.ValidCci

/**
 * Fake determinista para testear ViewModels en la JVM, sin `.so` y sin emulador.
 *
 * **No calcula nada**: devuelve lo que se le configuró. Poner aritmética acá sería
 * reimplementar el core en Kotlin, que es lo contrario de lo que la POC demuestra. Los
 * valores que se le pasan salen de `contracts/cases.json`.
 */
class FakeCoreFinanciero(
    private var nextAdd: Result<String> = Result.success("0.30"),
    private var nextSubtract: Result<String> = Result.success("0.00"),
    private var nextItf: Result<String> = Result.success("0.05"),
    private var nextCci: Result<ValidCci> = Result.success(ValidCci("002", "Banco Demo Uno", "191", "001234567890")),
    private var nextCard: Result<ValidCard> = Result.success(ValidCard("Visa", "4111 **** **** 1111")),
    private var nextEncrypt: Result<String> = Result.success("bdca3931"),
    private var nextDecrypt: Result<String> = Result.success("4111111111111111"),
    private var nextTransfer: Result<TransferResult> = Result.success(
        TransferResult(emptyList(), "0.05", "100.05", "TRF-0001", 120u),
    ),
    private val version: String = "1.0.0+test",
) : CoreFinanciero {
    override fun add(a: String, b: String) = nextAdd
    override fun subtract(a: String, b: String) = nextSubtract
    override fun calculateItf(amount: String) = nextItf
    override fun validateCci(cci: String) = nextCci
    override fun validateCard(number: String) = nextCard
    override fun encrypt(text: String, keyHex: String, nonceHex: String) = nextEncrypt
    override fun decrypt(ciphertextHex: String, keyHex: String, nonceHex: String) = nextDecrypt
    override fun transfer(accounts: List<Account>, request: TransferRequest) = nextTransfer
    override fun coreVersion() = version

    fun failNextTransfer(e: Throwable) { nextTransfer = Result.failure(e) }
    fun failNextAdd(e: Throwable) { nextAdd = Result.failure(e) }
    fun failNextCard(e: Throwable) { nextCard = Result.failure(e) }
}

File: apps/ios/ios-rust-testTests/FakeCoreFinanciero.swift
@testable import ios_rust_test

/// Determinista y sin núcleo: los tests de ViewModel prueban el ViewModel, no el core.
/// Quien prueba el core es `ContractTest`, contra los 28 casos reales.
///
/// Cada closure tiene un valor por defecto, así que un test solo reemplaza lo que le
/// importa.
final class FakeCoreFinanciero: CoreFinanciero {
    var addResult: (String, String) throws -> String = { a, b in "\(a)+\(b)" }
    var subtractResult: (String, String) throws -> String = { a, b in "\(a)-\(b)" }
    var itfResult: (String) throws -> String = { _ in "0.01" }
    var cciResult: (String) throws -> ValidCci = { _ in
        ValidCci(
            bankCode: "002", bankName: "Banco Demo Uno", branch: "191", account: "001234567890")
    }
    var cardResult: (String) throws -> ValidCard = { _ in
        ValidCard(brand: "Visa", masked: "4111 **** **** 1111")
    }
    var encryptResult: (String) throws -> String = { _ in "deadbeef" }
    var decryptResult: (String) throws -> String = { _ in "4111111111111111" }
    var transferResult: ([Account], TransferRequest) throws -> TransferResult = { accounts, _ in
        TransferResult(
            accounts: accounts,
            itfFee: "0.01",
            totalDebited: "100.01",
            receipt: "TRF-9047-1065-10000",
            simulatedLatencyMs: 0  // cero: los tests no esperan de verdad
        )
    }
    var version = "1.0.0+test"

    func add(a: String, b: String) throws -> String { try addResult(a, b) }
    func subtract(a: String, b: String) throws -> String { try subtractResult(a, b) }
    func calculateItf(amount: String) throws -> String { try itfResult(amount) }
    func validateCci(cci: String) throws -> ValidCci { try cciResult(cci) }
    func validateCard(number: String) throws -> ValidCard { try cardResult(number) }
    func encrypt(text: String, keyHex: String, nonceHex: String) throws -> String {
        try encryptResult(text)
    }
    func decrypt(ciphertextHex: String, keyHex: String, nonceHex: String) throws -> String {
        try decryptResult(ciphertextHex)
    }
    func transfer(accounts: [Account], request: TransferRequest) throws -> TransferResult {
        try transferResult(accounts, request)
    }
    func coreVersion() -> String { version }
}
